src/clj/rems/workflow/dynamic.clj

```diff
@@ -966,6 +1046,73 @@
+        injections {:valid-user? #{"somebody" "somebody2" "applicant"}}]
+    (testing "invitation token is available before use"
+      (is (= ["very-secure"]
+             (keys (:invitation-tokens application)))))
+
+    (testing "invitation token is not available after use"
+      (is (empty?
+           (keys (:invitation-tokens
+                  (apply-commands application
+                                  [{:type ::accept-invitation :actor "somebody" :token "very-secure"}]
+                                  injections))))))
+
+    (testing "invited member can join draft"
+      (is (= [{:userid "applicant"} {:userid "somebody"}]
+             (:members
+              (apply-commands application
+                              [{:type ::accept-invitation :actor "somebody" :token "very-secure"}]
+                              injections)))))
+
+    (let [application (apply-events application
+                                    [{:event/type :application.event/member-added
+                                      :event/actor "applicant"
+                                      :application/member {:userid "somebody"}}])]
+      (testing "invited member can't join if they are already a member"
+        (is (= {:errors [{:type :already-member :application-id (:id application)}]}
+               (:result (catch-ex
+                         (apply-command application
+                                        {:type ::accept-invitation :actor "somebody" :token "very-secure"}
+                                        injections)))))))
+
+    (testing "invalid token can't be used to join"
+      (is (= {:errors [{:type :t.actions.errors/invalid-token :token "wrong-token"}]}
+             (:result
+              (catch-ex
+               (apply-commands application
+                               [{:type ::accept-invitation :actor "somebody" :token "wrong-token"}]
+                               injections))))))
+
+    (testing "token can't be used twice"
+      (is (= {:errors [{:type :t.actions.errors/invalid-token :token "very-secure"}]}
+             (:result
+              (catch-ex
+               (apply-commands application
+                               [{:type ::accept-invitation :actor "somebody" :token "very-secure"}
+                                {:type ::accept-invitation :actor "somebody2" :token "very-secure"}]
+                               injections))))))
+
+    (let [submitted (apply-events application
+                                  [{:event/type :application.event/submitted
+                                    :event/actor "applicant"}])]
+      (testing "invited member can join submitted application"
+        (is (= [{:userid "applicant"} {:userid "somebody"}]
+               (:members
+                (apply-commands application
+                                [{:type ::accept-invitation :actor "somebody" :token "very-secure"}]
+                                injections))))))))
```

Comment:
how about an accepted application?

Reply:
I didn't intend to test all the possible states where it's possible to join. It's more significant at least that a submitted application can be joined since a draft is already checked. Should I test all the states here?

Reply:
I'd like at least one negative case where the state of the application is the reason for the failure (instead of invalid code or missing invitation), but I'm not insisting.

Reply:
Well accepting the invitation should always be possible but maybe not when the application has been closed already. That also creates a new situation that has to be shown to the joiner so maybe I'll add that one.